### PR TITLE
fix. Prevent character deletion if you have an email

### PR DESCRIPTION
**Pull request by pangolp**
## Changes Proposed:

This PR proposes changes to:
-  [x] Core (units, players, creatures, game systems).

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/18741

## Tests Performed:

This PR has been:
- [x] Tested in-game by the author.

## How to Test the Changes:

When trying to delete a character, which has a pending mail, the deletion is cancelled.

1. Create 2 characters
2. Send a collect item to another character
3. Attempt to delete the character that should have the item.
4. The character cannot be deleted, because he has a pending mail.

https://github.com/azerothcore/azerothcore-wotlk/assets/2810187/8a8a01c8-5243-466e-9818-7c65d3113d77

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Handlers/CharacterHandler.cpp [hunk 1]
```diff
@@ -635,11 +635,20 @@ void WorldSession::HandleCharDeleteOpcode(WorldPacket& recvData)
         return;
     }
 
-    if (CharacterCacheEntry const* playerData = sCharacterCache->GetCharacterCacheByGuid(guid))
+    CharacterCacheEntry const* playerData = sCharacterCache->GetCharacterCacheByGuid(guid);
+
+    if (playerData)
```
Comment (Nyeriah) [inline src/server/game/Handlers/CharacterHandler.cpp:640]: you don't need to change this

Comment (pangolp) [reply src/server/game/Handlers/CharacterHandler.cpp:640]: Ok, now I'll reverse it then.

**Comment by PkllonG** [comment]
Tested

**Comment by PkllonG** [comment]
Email recipients cannot be deleted

**Comment by PkllonG** [comment]
Can you prompt this?
SendCharFactionChange(CHAR_CREATE_CHARACTER_DELETE_MAIL, factionChangeInfo.get());

**Comment by pangolp** [comment]
> Can you prompt this?
> SendCharFactionChange(CHAR_CREATE_CHARACTER_DELETE_MAIL, factionChangeInfo.get());

I was thinking about being able to do that. Maybe it's the best.

**Comment by PkllonG** [comment]
```
        // check mailbox
        if (playerData->MailCount)
        {
            //SendCharDelete(CHAR_DELETE_FAILED);
            std::shared_ptr<CharacterFactionChangeInfo> factionChangeInfo = std::make_shared<CharacterFactionChangeInfo>();
            SendCharFactionChange(CHAR_CREATE_CHARACTER_DELETE_MAIL, factionChangeInfo.get());
            return;
        }
```
I tried it and this disables the deletion, but the prompt is wrong

**Comment by pangolp** [comment]
> ```
>         // check mailbox
>         if (playerData->MailCount)
>         {
>             //SendCharDelete(CHAR_DELETE_FAILED);
>             std::shared_ptr<CharacterFactionChangeInfo> factionChangeInfo = std::make_shared<CharacterFactionChangeInfo>();
>             SendCharFactionChange(CHAR_CREATE_CHARACTER_DELETE_MAIL, factionChangeInfo.get());
>             return;
>         }
> ```
> I tried it and this disables the deletion, but the prompt is wrong

I'll try to change it in a while. I didn't want to mix things up, but yeah. That message would be more representative of the error.

**Comment by heyitsbench** [comment]
Classic for the record seems to use a message that doesn't exist on 3.3.5. I would recommend staying with `CHAR_DELETE_FAILED` until the message can be verified against a 3.3.5 source.

**Comment by PkllonG** [comment]
I looked up the client file and the client doesn't have a prompt error for the PR.

**Comment by PkllonG** [comment]
> Classic for the record seems to use a message that doesn't exist on 3.3.5. I would recommend staying with `CHAR_DELETE_FAILED` until the message can be verified against a 3.3.5 source.

agree

**Comment by PkllonG** [comment]
You're working so fast......

**Comment by pangolp** [comment]
They're going to drive me crazy haha Okay. Then I change it again

**Comment by PkllonG** [comment]
Please wait

**Comment by PkllonG** [comment]
I'll try to see if there's any other way to disable it

**Comment by PkllonG** [comment]
Sorry to disrupt your work..

**Comment by pangolp** [comment]
> Sorry to disrupt your work..
> 

No problem. Then I adjust it.

**Comment by sudlud** [comment]
Whatever the final solution will be, I'd really prefer a solution which clearly communicates to the player why the character can't be deleted (if possible)

Otherwise this might lead to confusion maybe

**Review by Nyeriah — APPROVED** [review]

**Pull request merged**